<!-- Launch file for starting the openai Panda simulator -->
<launch>

    <!--## ARGUMENTS ##-->
    <!-- Simulation arguments -->
    <arg name="paused" default="false" />
    <arg name="world_name" default="/$(find panda_openai_sim)/resources/worlds/grasp_training.world" />
    <arg name="rviz" default="true" />
    <arg name="rviz_file" default="$(find panda_openai_sim)/launch/panda.rviz" />
    <arg name="moveit" default="true" />
    <arg name="moveit_rviz_file" default="$(find panda_openai_sim)/launch/panda.rviz" />
    <arg name="load_gripper" default="true" />
    <arg name="grip_site" default="false" />
    <arg name="grip_site_xyz" default="0.0 0.0 0.10" />
    <arg name="grip_site_rpy" default="0.0 0.0 0.0" />
    <!-- Control arguments -->
    <arg name="use_group_controller" default="false" />
    <!-- Wheter to use group controllers for the panda control server -->
    <arg name="preload_controllers" default="true" />
    <!-- Wheter to preload all non essential controllers -->

    <!-- ## LOAD ROBOT DESCRIPTION ## -->
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(find franka_description)/robots/panda_arm_hand.urdf.xacro" />

    <!--## LOAD CONTROLLERS ##
        # NOTE: Make sure the load_controllers argument is set to false if this block is uncommented -->
    <include file="$(find panda_openai_sim)/launch/controllers.launch">
        <arg name="load_gripper" value="$(arg load_gripper)" />
        <arg name="preload_controllers" value="$(arg preload_controllers)" />
    </include>

    <!--## LOAD MOVEIT CONTROL SERVICES ##-->
    <node if="$(arg moveit)" pkg="panda_openai_sim" type="panda_moveit_server.py" name="panda_moveit_planner_server" required="true">
        <param name="create_all_services" value="false" />
    </node>
    <node pkg="panda_openai_sim" type="panda_control_server.py" name="panda_control_server" required="true">
        <param name="use_group_controller" value="$(arg use_group_controller)" />
        <param name="create_all_services" value="false" />
    </node>

    <!--## SIMULATION ##-->
    <!-- Launch simulation -->
    <include file="$(find panda_simulation)/launch/simulation.launch">
        <arg name="paused" value="$(arg paused)" />
        <arg name="rviz" value="$(arg rviz)" />
        <arg name="moveit" value="$(arg moveit)" />
        <arg name="world_name" value="$(arg world_name)" />
        <arg name="moveit_rviz_file" value="$(arg moveit_rviz_file)" />
        <arg name="load_gripper" value="$(arg load_gripper)" />
        <arg name="load_robot_state_publisher" value="true" />
        <arg name="load_controllers" value="false" />
        <arg name="grip_site" value="$(arg grip_site)" />
        <arg name="grip_site_xyz" value="$(arg grip_site_xyz)" />
        <arg name="grip_site_rpy" value="$(arg grip_site_rpy)" />
    </include>

    <!-- ## Increase Movit execution time limit (See issue #33)-->
    <!-- # FIXME: Check if this is the right way -->
    <param name="move_group/trajectory_execution/allowed_execution_duration_scaling" value="4.0" />
    <!-- <param name="move_group/trajectory_execution/execution_duration_monitoring" value="false" /> -->

    <!-- ## LOAD RVIZ ##
         # NOTE: Only moveit was not used -->
    <include if="$(eval not arg('moveit') and arg('rviz'))" file="$(find panda_openai_sim)/launch/panda_rviz.launch">
        <arg name="rviz_file" value="$(arg rviz_file)" />
    </include>
</launch>